Databricks test fail when using `run.sh` script

They pass in CI though.

```console
$ ./run.sh pytest tests/test_databricks_backend.py 
Creating cohort-extractor_app_run ... done
=========================================================== test session starts ============================================================
platform linux -- Python 3.8.0, pytest-5.4.2, py-1.10.0, pluggy-0.13.1
rootdir: /workspace, inifile: pytest.ini
plugins: freezegun-0.4.2
collected 1 item                                                                                                                           

tests/test_databricks_backend.py F                                                                                                   [100%]

================================================================= FAILURES =================================================================
____________________________________________________________ test_minimal_study ____________________________________________________________

    def test_minimal_study():
        session = make_session()
        patient_1 = Patient(
            DateOfBirth="1980-01-01",
            Sex="M",
            CodedEvents=[
                CodedEvent(CTV3Code="aaaaa", ConsultationDate="2018-01-01"),
                CodedEvent(CTV3Code="bbbbb", ConsultationDate="2020-01-01"),
            ],
        )
        patient_2 = Patient(
            DateOfBirth="1965-01-01",
            Sex="F",
            CodedEvents=[CodedEvent(CTV3Code="zzzzz", ConsultationDate="2018-01-01")],
        )
        session.add_all([patient_1, patient_2])
        session.commit()
    
        codes = codelist(["aaaaa", "bbbbb", "ccccc"], "ctv3")
    
>       study = StudyDefinition(
            population=patients.all(),
            sex=patients.sex(),
            age=patients.age_as_of("2020-01-01"),
            has_event=patients.with_these_clinical_events(codes),
        )

tests/test_databricks_backend.py:46: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cohortextractor/study_definition.py:34: in __init__
    self.backend = self.create_backend()
cohortextractor/study_definition.py:144: in create_backend
    return Backend(
cohortextractor/databricks_backend.py:7: in __init__
    self.session = self.get_spark_session(database_url)
cohortextractor/databricks_backend.py:86: in get_spark_session
    spark_context = SparkContext(conf=conf)
/usr/local/lib/python3.8/dist-packages/pyspark/context.py:144: in __init__
    SparkContext._ensure_initialized(self, gateway=gateway, conf=conf)
/usr/local/lib/python3.8/dist-packages/pyspark/context.py:331: in _ensure_initialized
    SparkContext._gateway = gateway or launch_gateway(conf)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

conf = <pyspark.conf.SparkConf object at 0x7f274331b700>
popen_kwargs = {'env': {'DATABASE_URL': 'jdbc:sqlserver://localhost:15785;databaseName=Test_OpenCorona', 'DATABRICKS_DATABASE_URL': '...ql:1433/Test_EMIS', ...}, 'preexec_fn': <function launch_gateway.<locals>.preexec_func at 0x7f2742ddc5e0>, 'stdin': -1}

    def launch_gateway(conf=None, popen_kwargs=None):
        """
        launch jvm gateway
    
        Parameters
        ----------
        conf : :py:class:`pyspark.SparkConf`
            spark configuration passed to spark-submit
        popen_kwargs : dict
            Dictionary of kwargs to pass to Popen when spawning
            the py4j JVM. This is a developer feature intended for use in
            customizing how pyspark interacts with the py4j JVM (e.g., capturing
            stdout/stderr).
    
        Returns
        -------
        ClientServer or JavaGateway
        """
        if "PYSPARK_GATEWAY_PORT" in os.environ:
            gateway_port = int(os.environ["PYSPARK_GATEWAY_PORT"])
            gateway_secret = os.environ["PYSPARK_GATEWAY_SECRET"]
            # Process already exists
            proc = None
        else:
            SPARK_HOME = _find_spark_home()
            # Launch the Py4j gateway using Spark's run command so that we pick up the
            # proper classpath and settings from spark-env.sh
            on_windows = platform.system() == "Windows"
            script = "./bin/spark-submit.cmd" if on_windows else "./bin/spark-submit"
            command = [os.path.join(SPARK_HOME, script)]
            if conf:
                for k, v in conf.getAll():
                    command += ['--conf', '%s=%s' % (k, v)]
            submit_args = os.environ.get("PYSPARK_SUBMIT_ARGS", "pyspark-shell")
            if os.environ.get("SPARK_TESTING"):
                submit_args = ' '.join([
                    "--conf spark.ui.enabled=false",
                    submit_args
                ])
            command = command + shlex.split(submit_args)
    
            # Create a temporary directory where the gateway server should write the connection
            # information.
            conn_info_dir = tempfile.mkdtemp()
            try:
                fd, conn_info_file = tempfile.mkstemp(dir=conn_info_dir)
                os.close(fd)
                os.unlink(conn_info_file)
    
                env = dict(os.environ)
                env["_PYSPARK_DRIVER_CONN_INFO_PATH"] = conn_info_file
    
                # Launch the Java gateway.
                popen_kwargs = {} if popen_kwargs is None else popen_kwargs
                # We open a pipe to stdin so that the Java gateway can die when the pipe is broken
                popen_kwargs['stdin'] = PIPE
                # We always set the necessary environment variables.
                popen_kwargs['env'] = env
                if not on_windows:
                    # Don't send ctrl-c / SIGINT to the Java gateway:
                    def preexec_func():
                        signal.signal(signal.SIGINT, signal.SIG_IGN)
                    popen_kwargs['preexec_fn'] = preexec_func
                    proc = Popen(command, **popen_kwargs)
                else:
                    # preexec_fn not supported on Windows
                    proc = Popen(command, **popen_kwargs)
    
                # Wait for the file to appear, or for the process to exit, whichever happens first.
                while not proc.poll() and not os.path.isfile(conn_info_file):
                    time.sleep(0.1)
    
                if not os.path.isfile(conn_info_file):
>                   raise Exception("Java gateway process exited before sending its port number")
E                   Exception: Java gateway process exited before sending its port number

/usr/local/lib/python3.8/dist-packages/pyspark/java_gateway.py:108: Exception
----------------------------------------------------------- Captured stderr call -----------------------------------------------------------
JAVA_HOME is not set
========================================================= short test summary info ==========================================================
FAILED tests/test_databricks_backend.py::test_minimal_study - Exception: Java gateway process exited before sending its port number
============================================================ 1 failed in 1.55s =============================================================
ERROR: 1
```

--- a/cohortextractor/databricks_backend.py
+++ b/cohortextractor/databricks_backend.py
@@ -132,7 +132,7 @@ class DatabricksBackend:
         # as this connection is from one container to another, not from host
         # to container.
         database_url = os.environ.get(
-            "DATABRICKS_TEST_DATABASE_URL",
+            "DATABRICKS_DATASOURCE_DATABASE_URL",
             "jdbc:sqlserver://localhost:15785;databaseName=Test_OpenCorona",
         )
 

--- a/cohortextractor/databricks_backend.py
+++ b/cohortextractor/databricks_backend.py
@@ -1,3 +1,4 @@
+import os
 from datetime import datetime
 from functools import reduce
 from pathlib import Path
@@ -122,13 +123,23 @@ class DatabricksBackend:
         This currently connects to a SQL Server database.  It is expected that this will
         need to change for production!
         """
+        # The default JDBC URL works for tests when they are run directly on
+        # the host, with the database servers running in Docker Compose.
+        #
+        # If running the cohort-extractor tests *in* Docker Compose, e.g. as
+        # run.sh does, for the connection to succeed, this URL needs to be:
+        # jdbc:sqlserver://mssql:1433;databaseName=Test_OpenCorona
+        # as this connection is from one container to another, not from host
+        # to container.
+        database_url = os.environ.get(
+            "DATABRICKS_TEST_DATABASE_URL",
+            "jdbc:sqlserver://localhost:15785;databaseName=Test_OpenCorona",
+        )
 
         reader = (
             self.session.read.format("jdbc")
             .option("driver", "com.microsoft.sqlserver.jdbc.SQLServerDriver")
-            .option(
-                "url", "jdbc:sqlserver://localhost:15785;databaseName=Test_OpenCorona"
-            )
+            .option("url", database_url)
             .option("user", "SA")
             .option("password", "Your_password123!")
         )

--- a/run.sh
+++ b/run.sh
@@ -24,6 +24,6 @@ exec docker-compose run \
   -e TPP_DATABASE_URL='mssql://SA:Your_password123!@mssql:1433/Test_OpenCorona' \
   -e EMIS_DATABASE_URL=presto://presto:8080/mssql/dbo \
   -e EMIS_DATASOURCE_DATABASE_URL='mssql://SA:Your_password123!@mssql:1433/Test_EMIS' \
-  -e DATABRICKS_DATABASE_URL='jdbc:sqlserver://mssql:1433;databaseName=Test_OpenCorona' \
+  -e DATABRICKS_DATABASE_URL='databricks' \
   app \
   -- "$@"

--- a/run.sh
+++ b/run.sh
@@ -25,5 +25,6 @@ exec docker-compose run \
   -e EMIS_DATABASE_URL=presto://presto:8080/mssql/dbo \
   -e EMIS_DATASOURCE_DATABASE_URL='mssql://SA:Your_password123!@mssql:1433/Test_EMIS' \
   -e DATABRICKS_DATABASE_URL='databricks' \
+  -e DATABRICKS_TEST_DATABASE_URL='jdbc:sqlserver://mssql:1433;databaseName=Test_OpenCorona' \
   app \
   -- "$@"

--- a/run.sh
+++ b/run.sh
@@ -25,6 +25,6 @@ exec docker-compose run \
   -e EMIS_DATABASE_URL=presto://presto:8080/mssql/dbo \
   -e EMIS_DATASOURCE_DATABASE_URL='mssql://SA:Your_password123!@mssql:1433/Test_EMIS' \
   -e DATABRICKS_DATABASE_URL='databricks' \
-  -e DATABRICKS_TEST_DATABASE_URL='jdbc:sqlserver://mssql:1433;databaseName=Test_OpenCorona' \
+  -e DATABRICKS_DATASOURCE_DATABASE_URL='jdbc:sqlserver://mssql:1433;databaseName=Test_OpenCorona' \
   app \
   -- "$@"

